# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [193]:
# Your code here
import pandas as pd

url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
customer_df = pd.read_csv(url)
pd.options.display.float_format = '{:,.2f}'.format
customer_df

,Customer,ST,GENDER,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount
0,RB50392,Washington,NaN,Master,NaN,0.00,"1,000.00",1/0/00,Personal Auto,Four-Door Car,2.70
1,QZ44356,Arizona,F,Bachelor,697953.59%,0.00,94.00,1/0/00,Personal Auto,Four-Door Car,"1,131.46"
2,AI49188,Nevada,F,Bachelor,1288743.17%,"48,767.00",108.00,1/0/00,Personal Auto,Two-Door Car,566.47
3,WW63253,California,M,Bachelor,764586.18%,0.00,106.00,1/0/00,Corporate Auto,SUV,529.88
4,GA49547,Washington,M,High School or Below,536307.65%,"36,357.00",68.00,1/0/00,Personal Auto,Four-Door Car,17.27
...,...,...,...,...,...,...,...,...,...,...,...
4003,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4004,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4005,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4006,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [194]:
#Identify the dimensions of the dataset by determining the number of rows and columns it contains.
customer_df.shape



(4008, 11)

In [195]:
# Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. 

customer_df.dtypes



Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

In [196]:
#You should also provide suggestions for fixing any incorrect data types. Conviertiendo a numéricos y splitear compaints:
#Arreglar CLV:
customer_df["Customer Lifetime Value"] = customer_df["Customer Lifetime Value"].replace('%', "" , regex=True)
customer_df["Customer Lifetime Value"] = customer_df["Customer Lifetime Value"].astype(str).str.strip() #remove spaces 
customer_df["Customer Lifetime Value"] = pd.to_numeric(customer_df["Customer Lifetime Value"], errors='coerce').round(2) #numeric



#Separate 'Number of Open Complaints' into three distinct columns
customer_df[['Open Complaint 1', 'Open Complaint 2', 'Open Complaint 3']] = customer_df['Number of Open Complaints'].str.split('/', expand=True)
#Convert the new columns to numeric
customer_df[['Open Complaint 1', 'Open Complaint 2', 'Open Complaint 3']] = customer_df[['Open Complaint 1', 'Open Complaint 2', 'Open Complaint 3']].apply(pd.to_numeric, errors='coerce')
customer_df.dtypes

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value      float64
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
Open Complaint 1             float64
Open Complaint 2             float64
Open Complaint 3             float64
dtype: object

In [197]:
customer_df.head(10)

,Customer,ST,GENDER,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount,Open Complaint 1,Open Complaint 2,Open Complaint 3
0,RB50392,Washington,NaN,Master,NaN,0.00,"1,000.00",1/0/00,Personal Auto,Four-Door Car,2.70,1.00,0.00,0.00
1,QZ44356,Arizona,F,Bachelor,"697,953.59",0.00,94.00,1/0/00,Personal Auto,Four-Door Car,"1,131.46",1.00,0.00,0.00
2,AI49188,Nevada,F,Bachelor,"1,288,743.17","48,767.00",108.00,1/0/00,Personal Auto,Two-Door Car,566.47,1.00,0.00,0.00
3,WW63253,California,M,Bachelor,"764,586.18",0.00,106.00,1/0/00,Corporate Auto,SUV,529.88,1.00,0.00,0.00
4,GA49547,Washington,M,High School or Below,"536,307.65","36,357.00",68.00,1/0/00,Personal Auto,Four-Door Car,17.27,1.00,0.00,0.00
5,OC83172,Oregon,F,Bachelor,"825,629.78","62,902.00",69.00,1/0/00,Personal Auto,Two-Door Car,159.38,1.00,0.00,0.00
6,XZ87318,Oregon,F,College,"538,089.86","55,350.00",67.00,1/0/00,Corporate Auto,Four-Door Car,321.60,1.00,0.00,0.00
7,CF85061,Arizona,M,Master,"721,610.03",0.00,101.00,1/0/00,Corporate Auto,Four-Door Car,363.03,1.00,0.00,0.00
8,DY87989,Oregon,M,Bachelor,"2,412,750.40","14,072.00",71.00,1/0/00,Corporate Auto,Four-Door Car,511.20,1.00,0.00,0.00
9,BQ94931,Oregon,F,College,"738,817.81","28,812.00",93.00,1/0/00,Special Auto,Four-Door Car,425.53,1.00,0.00,0.00


In [198]:
# Identify the number of unique values for each column and determine which columns appear to be categorical. 
customer_df.nunique()



Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
Open Complaint 1                1
Open Complaint 2                6
Open Complaint 3                1
dtype: int64

In [199]:
# You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
# Get categorical and numeric column names from the dataframe

object_cols = ["Customer","ST","GENDER","Education","Policy Type","Vehicle Class"]
num_cols = ["Customer Lifetime Value","Income","Monthly Premium Auto","Open Complaint 1","Open Complaint 2","Open Complaint 3","Total Claim Amount"]

object_cols = customer_df.select_dtypes(include='object')
num_cols = customer_df.select_dtypes(include=['int64','float64'])

for col in object_cols:
    print (f"{col} unique values: {customer_df[col].unique()})")



Customer unique values: ['RB50392' 'QZ44356' 'AI49188' ... 'CW49887' 'MY31220' nan])
ST unique values: ['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Cali' 'AZ' 'WA'
 nan])
GENDER unique values: [nan 'F' 'M' 'Femal' 'Male' 'female'])
Education unique values: ['Master' 'Bachelor' 'High School or Below' 'College' 'Bachelors' 'Doctor'
 nan])
Number of Open Complaints unique values: ['1/0/00' '1/2/00' '1/1/00' '1/3/00' '1/5/00' '1/4/00' nan])
Policy Type unique values: ['Personal Auto' 'Corporate Auto' 'Special Auto' nan])
Vehicle Class unique values: ['Four-Door Car' 'Two-Door Car' 'SUV' 'Luxury SUV' 'Sports Car'
 'Luxury Car' nan])


In [200]:

#Arreglamos valores de estado
customer_df["ST"] = customer_df["ST"].replace('WA', 'Washington')
customer_df["ST"] = customer_df["ST"].replace('Cali', 'California')
customer_df["ST"] = customer_df["ST"].replace('AZ', 'Arizona')

# Arreglar valores de genero
customer_df["GENDER"] = customer_df["GENDER"].replace('female', 'F')
customer_df["GENDER"] = customer_df["GENDER"].replace('Femal', 'F')
customer_df["GENDER"] = customer_df["GENDER"].replace('Male', 'M')

#Arreglar education
customer_df["Education"] = customer_df["Education"].replace('Bachelors', 'Bachelor')


#print(customer_df['GENDER'].unique())
#customer_df.nunique()

for col in object_cols:
    print (f"{col} unique values: {customer_df[col].unique()})")


Customer unique values: ['RB50392' 'QZ44356' 'AI49188' ... 'CW49887' 'MY31220' nan])
ST unique values: ['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' nan])
GENDER unique values: [nan 'F' 'M'])
Education unique values: ['Master' 'Bachelor' 'High School or Below' 'College' 'Doctor' nan])
Number of Open Complaints unique values: ['1/0/00' '1/2/00' '1/1/00' '1/3/00' '1/5/00' '1/4/00' nan])
Policy Type unique values: ['Personal Auto' 'Corporate Auto' 'Special Auto' nan])
Vehicle Class unique values: ['Four-Door Car' 'Two-Door Car' 'SUV' 'Luxury SUV' 'Sports Car'
 'Luxury Car' nan])


In [201]:
# Numerical columns, and give your insights.
for col in num_cols:
    print (f"{col} range: {customer_df[col].min()} - {customer_df[col].max()})")

Customer Lifetime Value range: 200435.07 - 5816655.35)
Income range: 0.0 - 99960.0)
Monthly Premium Auto range: 61.0 - 35354.0)
Total Claim Amount range: 0.382107 - 2893.239678)
Open Complaint 1 range: 1.0 - 1.0)
Open Complaint 2 range: 0.0 - 5.0)
Open Complaint 3 range: 0.0 - 0.0)


In [202]:

# Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. 
num_cols.describe()



,Customer Lifetime Value,Income,Monthly Premium Auto,Total Claim Amount,Open Complaint 1,Open Complaint 2,Open Complaint 3
count,"1,068.00","1,071.00","1,071.00","1,071.00","1,071.00","1,071.00","1,071.00"
mean,"793,690.26","39,295.70",193.23,404.99,1.00,0.39,0.00
std,"643,478.41","30,469.43","1,601.19",293.03,0.00,0.88,0.00
min,"200,435.07",0.00,61.00,0.38,1.00,0.00,0.00
25%,"403,407.99","14,072.00",68.00,202.16,1.00,0.00,0.00
50%,"588,174.23","36,234.00",83.00,354.73,1.00,0.00,0.00
75%,"896,287.20","64,631.00",109.50,532.80,1.00,0.00,0.00
max,"5,816,655.35","99,960.00","35,354.00","2,893.24",1.00,5.00,0.00


# You should also provide your conclusions based on these summary statistics.

* **Customer Lifetime Value**: no tenemos datos en 3 en ocasiones, la media es de 794k
* **Income**: hay clientes que no dan ingresos frente a otros que sí que son más rentables.
* **Monthly Premium Auto**: 193 € de media, aunque la cifra min y max varía mucho, de ahí que el std sea elevado.
* **Total Claim Amount**: 405 € de media, aunque la cifra min y max varía mucho, de ahí que el std sea elevado.
* **Open Complaint 1**: la gente tiene como máximo un parte	de media el primer año.
* **Open Complaint 2**:	la media baja porque puede haber outliers.
* **Open Complaint 3**: media es 0 porque no parece haber quejas.

In [204]:
# Compute summary statistics for categorical columns and providing your conclusions based on these statistics.
object_cols.describe()

,Customer,ST,GENDER,Education,Number of Open Complaints,Policy Type,Vehicle Class
count,1071,1071,954,1071,1071,1071,1071
unique,1071,8,5,6,6,3,6
top,RB50392,Oregon,F,Bachelor,1/0/00,Personal Auto,Four-Door Car
freq,1,320,457,324,830,780,576


## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [205]:
# The marketing team wants to know the top 5 less common customer locations.
#  Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common 
# locations in ascending order.

location_counts = customer_df['ST'].value_counts(ascending=True)
top_5_less_common_locations = location_counts.head(5)       
top_5_less_common_locations


ST
Nevada         98
Washington    111
Arizona       211
Oregon        320
California    331
Name: count, dtype: int64

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [206]:
# Your code here

#The sales team wants to know the total number of policies sold for each type of policy. 
# Create a pandas Series object that contains the policy types and their total number of policies sold, and 
# then retrieve the policy type with the highest number of policies sold.

policy_counts = customer_df['Policy Type'].value_counts()
top_policy_type = policy_counts.idxmax()
top_policy_type


'Personal Auto'

### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [207]:
# The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

personal_auto_income = customer_df[customer_df['Policy Type'] == 'Personal Auto']['Income'].mean()
corporate_auto_income = customer_df[customer_df['Policy Type'] == 'Corporate Auto']['Income'].mean()

personal_auto_income, corporate_auto_income

(np.float64(38180.69871794872), np.float64(41390.31196581197))

### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [208]:
# Your code here